package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,333 @@
+  *sender_id = strtoul(sbp_sender_id_string, NULL, 10);
+  return NO_ERROR;
+}
+
+RC get_config(skylark_client_config_t* config)
+{
+  RC rc = NO_ERROR;
+  if ((rc = get_broker_endpoint(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_accept_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_content_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  strcpy(config->encoding, STREAM_ENCODING);
+  if ((rc = get_device_uuid(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = format_device_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_sbp_sender_id(&config->sbp_sender_id)) < NO_ERROR) {
+    return rc;
+  }
+  return rc;
+}
+
+/**
+ *  Download (i.e., rover) processes.
+ */
+
+// Callback used by libcurl to pass data read from Skylark. Writes to pipe.
+//
+// TODO (mookerji): replace
+static size_t download_callback(void *p, size_t size, size_t n, void *up)
+{
+  int *fd = (int *)up;
+  ssize_t m = write(*fd, p, size * n);
+  return m;
+}
+
+RC download_process(skylark_client_config_t* config) {
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    exit(EXIT_FAILURE);
+  }
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_setopt(curl, CURLOPT_URL, config->endpoint_url);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &download_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &config->fd_out);
+#ifdef VERBOSE
+  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+#endif
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+  struct curl_slist *chunk = NULL;
+  chunk = curl_slist_append(chunk, config->encoding);
+  chunk = curl_slist_append(chunk, config->accept_type_header);
+  chunk = curl_slist_append(chunk, config->device_header);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+  res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
+    log_error(curl_easy_strerror(res));
+    curl_easy_cleanup(curl);
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_cleanup(curl);
+  curl_global_cleanup();
+  return NO_ERROR;
+}
+
+static void base_obs_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_OBS, len, msg);
+}
+
+static void base_llh_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_LLH, len, msg);
+}
+
+static void base_ecef_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_ECEF, len, msg);
+}
+
+u32 msg_read(u8 *buf, u32 n, void *context)
+{
+  int *fd = (int *)context;
+  ssize_t m = read(*fd, buf, n);
+  return m;
+}
+
+RC download_io_loop(skylark_client_config_t *config)
```

Comment:
I think I would move all of this machinery out of libskylark into the download task.